cpp/src/arrow/chunked_array.cc

```diff
@@ -43,15 +43,17 @@ class MemoryPool;
 // ChunkedArray methods
 
 ChunkedArray::ChunkedArray(ArrayVector chunks, std::shared_ptr<DataType> type)
-    : chunks_(std::move(chunks)), type_(std::move(type)) {
-  length_ = 0;
-  null_count_ = 0;
-
+    : chunks_(std::move(chunks)),
+      type_(std::move(type)),
+      resolver_{internal::ChunkResolver::FromChunks(chunks_)} {
```

Comment:
Constructing the `ChunkResolver` adds overhead to all `ChunkedArray` instances, and the `resolver_` is only used for `GetScalar()`. Maybe a better approach is to have a flag to control single `resolver_` initialization when calling `GetScalar`.
cc @pitrou